crates/oxc_diagnostics/src/lib.rs

```diff
@@ -35,3 +39,64 @@ pub struct MinifiedFileError(pub PathBuf);
 #[error("Failed to open file {0:?} with error \"{1}\"")]
 #[diagnostic(help("Failed to open file {0:?} with error \"{1}\""))]
 pub struct FailedToOpenFileError(pub PathBuf, pub std::io::Error);
+
+#[derive(Debug, Default, Error)]
+pub struct OxcDiagnostic {
+    message: Cow<'static, str>,
+    labels: Option<Vec<LabeledSpan>>,
+    help: Option<&'static str>,
+    severity: Option<Severity>,
+}
+
+impl OxcDiagnostic {
+    #[must_use]
+    pub fn new<T: Into<Cow<'static, str>>>(message: T) -> Self {
+        let message = message.into();
+        Self { message, labels: None, help: None, severity: None }
+    }
+
+    #[must_use]
+    pub fn label<T: Into<LabeledSpan>>(mut self, span: T) -> Self {
+        let span = span.into();
+        if let Some(labels) = &mut self.labels {
+            labels.push(span);
+        } else {
+            self.labels.replace(vec![span]);
+        }
+        self
+    }
+
+    #[must_use]
+    pub fn help(mut self, help: &'static str) -> Self {
+        self.help.replace(help);
+        self
+    }
+
+    #[must_use]
+    pub fn severity(mut self, severity: Severity) -> Self {
+        self.severity.replace(severity);
+        self
+    }
+}
+
+impl Display for OxcDiagnostic {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.message)
+    }
+}
+
+impl Diagnostic for OxcDiagnostic {
+    fn severity(&self) -> Option<Severity> {
+        Some(self.severity.unwrap_or(Severity::Warning))
```

Comment:
```suggestion
        self.severity
```

/// If `None`, reporters should treat this as [`Severity::Error`].